src/metabase/lib/stage.cljc

```diff
@@ -221,11 +221,23 @@
     unique-name-fn :- fn?]
    (or
     (existing-stage-metadata query stage-number)
-    (let [query (ensure-previous-stages-have-metadata query stage-number)]
+    (let [query (ensure-previous-stages-have-metadata query stage-number)
+          summary-cols (summary-columns query stage-number unique-name-fn)
+          field-cols (fields-columns query stage-number unique-name-fn)]
       ;; ... then calculate metadata for this stage
-      (or
-       (breakout-ags-fields-columns query stage-number unique-name-fn)
-       (lib.metadata.calculation/default-columns query stage-number (lib.util/query-stage query stage-number) unique-name-fn))))))
+      (cond
+        summary-cols
+        (into summary-cols field-cols)
+
+        field-cols
+        (do (doall field-cols)          ; force generation of unique names before join columns
```

Comment:
This is quite ugly, but probably inherent in the nature of `unique-name-fn`.
(I don't like this side-effectful name generation. I would much more prefer a solution where we a global view on what columns we are generating the names for.)

Reply:
Would it be possible to fill in `desired-column-alias` after metadata is gathered as a final step in `stage-metadata` etc? It doesn't seem like `unique-name-generator` is created in too many places.

Reply:
I'll open an issue for this and see what @camsaul thinks.